internal/logsink/logsink.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package logsink
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/juju/loggo/v2"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/logger"
+)
+
+const (
+	stateFlushed = "flushed"
+	stateTicked  = "ticked"
+)
+
+// NewWriterFunc is a function that creates a new writer.
+type NewWriterFunc func() (io.WriteCloser, error)
+
+// LogSink is a loggo.Writer that writes log messages to a file.
+type LogSink struct {
+	tomb           tomb.Tomb
+	internalStates chan string
+
+	writer io.Writer
+
+	batchSize     int
+	flushInterval time.Duration
+
+	inLogEntry   chan loggo.Entry
+	inLogRecords chan []logger.LogRecord
+
+	out chan []logRecord
+}
+
+// NewLogSink creates a new log sink that writes log messages to a file. There
+// can only be one writer writing to the same file at a time, otherwise bytes
+// will be written to the file in an interleaved manner (junk data). LogSink
+// writer will write log messages as JSON objects, one per line, even if the log
+// message is multiline. The batchSize parameter specifies the minimum number of
+// log messages to batch before writing to the underlying writer. The number of
+// entires can far exceed the batchSize if the log messages are large.
+func NewLogSink(writer io.Writer, batchSize int, flushInterval time.Duration) *LogSink {
+	return newLogSink(writer, batchSize, flushInterval, nil)
+}
+
+// newLogSink creates a new log sink that writes log messages to a file.
+func newLogSink(writer io.Writer, batchSize int, flushInterval time.Duration, internalStates chan string) *LogSink {
+	w := &LogSink{
+		internalStates: internalStates,
+
+		writer: writer,
+
+		batchSize:     batchSize,
+		flushInterval: flushInterval,
+
+		inLogEntry:   make(chan loggo.Entry),
+		inLogRecords: make(chan []logger.LogRecord),
+
+		out: make(chan []logRecord),
+	}
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Write sends a new log message to the writer.
+// This implements the loggo.Writer interface.
+func (w *LogSink) Write(entry loggo.Entry) {
+	select {
+	case <-w.tomb.Dying():
+		return
+	case w.inLogEntry <- entry:
+	}
+}
+
+// Log writes the given log records to the logger's storage.
+func (w *LogSink) Log(records []logger.LogRecord) error {
+	select {
+	case <-w.tomb.Dying():
+		return tomb.ErrDying
+	case w.inLogRecords <- records:
+		return nil
+	}
+}
+
+// Kill stops the writer.
+func (w *LogSink) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait blocks until the writer has stopped.
+func (w *LogSink) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *LogSink) loop() error {
+	w.tomb.Go(func() error {
+		buffer := new(bytes.Buffer)
```

Comment:
I am concerned about ordering here for the write, perhaps we should ensure this go routine has an exclusive lock until it exits, such that logs in the main goroutine that happened-after the logs being written here appear in the file in the correct order.

Reply:
There is no guarantee of correct ordering anyway, just because of the sources. One is from the controller, not via the API server, and the other is all the units from the API server. They will be out of order. It's up to the read side to attempt to order these. 

It would be better if we stored as parquet files, [which makes ingestion fast, but sorting on columns easier](https://15721.courses.cs.cmu.edu/spring2016/papers/abadi-sigmod2006.pdf). The issue then becomes easy access for general use cases (grep etc). At this point, we're just making a logging app inside of juju.